### PR TITLE
Homework 3 task 3 done

### PR DESCRIPTION
**Pull request by ivanzuev78**

### CODE REVIEW

File: HomeWork_3/task03/task03.py [hunk 1]
```diff
@@ -7,41 +8,43 @@ class Filter:
     functions that return True if object in list conforms to some criteria
     """
 
-    def __init__(self, functions):
+    def __init__(self, *functions: Callable):
         self.functions = functions
 
-    def apply(self, data):
-        return [item for item in data if all(i(item) for i in self.functions)]
+    def apply(self, data: List[dict]) -> List[dict]:
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/task03.py:14]: Why `List[dict]`?

File: HomeWork_3/task03/task03.py [hunk 2]
```diff
@@ -7,41 +8,43 @@ class Filter:
...
     """
-    filter_funcs = []
-    for key, value in keywords.items():
 
-        def keyword_filter_func(value):
-            return value[key] == value
+    def keyword_filter_func(key_to_check: Hashable):
+        def wrapper(input_dict: dict) -> bool:
+
+            if (
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/task03.py:31]: replace if to return

File: HomeWork_3/task03/task03.py [hunk 3]
```diff
@@ -7,41 +8,43 @@ class Filter:
...
+            else:
+                return False
+
+        return wrapper
+
+    filter_funcs = []
 
-        filter_funcs.append(keyword_filter_func)
-    return Filter(filter_funcs)
+    for key in keywords:
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/task03.py:43]: what type of the `key`?  Does this name good for it?

File: HomeWork_3/task03/test_task_3.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+from HomeWork_3.task03.task03 import make_filter, Filter
+
+
+def test_Filter():
+
+    assert Filter(
+        lambda a: a % 2 == 0, lambda a: a > 0, lambda a: isinstance(a, int)
+    ).apply(range(-10, 10)) == [2, 4, 6, 8]
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/test_task_3.py:8]: `range(-10, 10)`  just hardcode some numbers here.

File: HomeWork_3/task03/test_task_3.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def test_specified_filter_example():
+    sample_data = [
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/test_task_3.py:12]: Move sample to fixture.

Create two separate tests.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_3/task03/task03.py (hunk 1 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/task03.py:14]: Because I thought about example, not about function. 
Fixed.

File: HomeWork_3/task03/task03.py (hunk 2 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/task03.py:31]: Done

File: HomeWork_3/task03/task03.py (hunk 3 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/task03.py:43]: type(key) is str
Renamed

File: HomeWork_3/task03/test_task_3.py (hunk 5 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/test_task_3.py:12]: Done

File: HomeWork_3/task03/test_task_3.py (hunk 4 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/test_task_3.py:8]: Ok

File: HomeWork_3/task03/task03.py [hunk 6]
```diff
@@ -7,41 +8,40 @@ class Filter:
     functions that return True if object in list conforms to some criteria
     """
 
-    def __init__(self, functions):
+    def __init__(self, *functions: Callable):
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/task03.py:11]: When you change signature of the method, don't forget to check if docstring is still actual.

File: HomeWork_3/task03/task03.py [hunk 7]
```diff
@@ -7,41 +8,40 @@ class Filter:
...
+                else False
+            )
+
+        return wrapper
+
+    filter_funcs = []
 
-        filter_funcs.append(keyword_filter_func)
-    return Filter(filter_funcs)
+    for keyword in keywords:
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/task03.py:40]: Can you use list comprehension here?

File: HomeWork_3/task03/test_task_3.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    input = [
+        {
+            "name": "Bill",
+            "last_name": "Gilbert",
+            "occupation": "was here",
+            "type": "person",
+        },
+        {"is_dead": True, "kind": "parrot", "type": "bird", "name": "polly"},
+    ]
+    yield input
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/test_task_3.py:24]: Why do you need yield here?

File: HomeWork_3/task03/test_task_3.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def test_specified_filter_example(sample_data):
+
+    assert make_filter(name="polly", type="bird").apply(sample_data) == [sample_data[1]]
+
+
+@pytest.fixture
+def specified_filter_input():
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/test_task_3.py:33]: Could you rename sample_data and specified_filter_input so I can understand the difference between them?

File: HomeWork_3/task03/test_task_3.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            "name": "Bill",
+            "last_name": "Paper",
+        },
+        {"name": "Bill", "last_name": "Wodoomagic", "type": "person"},
+    ]
+    yield input
+
+
+def test_specified_filter__no_one_in_input_are_good(specified_filter_input):
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/test_task_3.py:50]: `__` -> `_`

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_3/task03/task03.py (hunk 6 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/task03.py:11]: I will try to remember.

File: HomeWork_3/task03/test_task_3.py (hunk 10 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/test_task_3.py]: Done

File: HomeWork_3/task03/test_task_3.py (hunk 9 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/test_task_3.py]: Yes, of course.

File: HomeWork_3/task03/test_task_3.py [hunk 11]
```diff
@@ -44,16 +46,17 @@ def specified_filter_input():
         },
         {"name": "Bill", "last_name": "Wodoomagic", "type": "person"},
     ]
-    yield input
+    return input_
 
 
-def test_specified_filter__no_one_in_input_are_good(specified_filter_input):
+def test_specified_filter_no_one_in_input_are_good(some_data_to_test_filter):
```
Comment (Cjkjvfnby) [inline HomeWork_3/task03/test_task_3.py:52]: `some_data_to_test_filter`  `data_from_example_of_the_task`   both names are abstract and both of them does not describe the content.    If next person decides to add one more test which of the should be used?

Maybe you dont need a fixtures?  Just put samples to each test.   You don't need good name inside the test, and each test can use own set of data.   Actually it is not a duplication, becasue if you change data inside one test, you don't need to change another.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_3/task03/task03.py [hunk 12]
```diff
@@ -1,47 +1,44 @@
 # I decided to write a code that generates data filtering object from a list of keyword parameters:
+from typing import Callable, List, Hashable, Optional, Any, Iterable
 
 
 class Filter:
     """
-    Helper filter class. Accepts a list of single-argument
+    Helper filter class. Accepts single-argument
```
[the remark this answers is not in the document]
Comment (ivanzuev78) [reply HomeWork_3/task03/task03.py:7]: Not single-argument, but single-argument functions.

File: HomeWork_3/task03/test_task_3.py (hunk 11 above)
Comment (ivanzuev78) [reply HomeWork_3/task03/test_task_3.py:52]: Very useful note.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: HomeWork_3/task03/test_task_3.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            "name": "Bill",
+            "last_name": "Paper",
+        },
+        {"name": "Bill", "last_name": "Wodoomagic", "type": "person"},
+    ]
+
+    assert make_filter(name="Bill", type="bird").apply(input_data) == []
+
+
+def test_specified_filter_all_in_input_are_good():
```
[the remark this answers is not in the document]
Comment (ivanzuev78) [reply HomeWork_3/task03/test_task_3.py:52]: Test added

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**